lib/motsu-proc/src/default_storage_layout.rs

```diff
@@ -0,0 +1,86 @@
+//! Defines the `#[derive(motsu::DefaultStorageLayout)]` procedural macro.
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{Data, DeriveInput, PathArguments, Type};
+
+pub fn impl_default_storage_layout(ast: &DeriveInput) -> TokenStream {
+    let name = &ast.ident;
+
+    let Data::Struct(ref data_struct) = ast.data else {
+        error!(ast, "DefaultStorageLayout can only be derived for structs");
+    };
+
+    let fields = match &data_struct.fields {
+        syn::Fields::Named(fields) => &fields.named,
+        syn::Fields::Unnamed(fields) => &fields.unnamed,
+        syn::Fields::Unit => return TokenStream::new(),
+    };
+
+    let mut field_initializations = Vec::new();
+
+    for field in fields {
+        let field_name = &field.ident;
+        let field_ty = &field.ty;
+
+        let Type::Path(type_path) = field_ty else {
+            error!(
+                field_ty,
+                "unsupported field type. Only path types are supported"
+            );
+        };
+
+        // Types when using `sol_storage!` look like this:
+        // `stylus_sdk::storage::type<generic arguments>`
+        // (e.g. uint256 is stylus_sdk::storage::StorageUint<256,4>).
+        // So we must first get the last argument, which is the main
+        // type.
+        let segments = &type_path.path.segments;
+        let last_segment = if segments.len() >= 3 {
+            segments.last().unwrap()
+        } else {
+            error!(type_path, "unexpected type path");
+        };
+        let main_type = &last_segment.ident;
+
+        // If the type has generic arguments form the token stream that
+        // we latter append to access `new` and `SLOT_BYTES`
+        let generic_args = match &last_segment.arguments {
+            PathArguments::AngleBracketed(args) => quote! { ::#args },
+            _ => quote! {},
+        };
+
+        // Reconstruct the type with the correct formatting
+        let type_ident =
+            quote! { stylus_sdk::storage:: #main_type #generic_args };
+
+        let field_init = quote! {
+            {
+                if offset + #type_ident::SLOT_BYTES as u8 > 32 {
+                    next_slot += 32;
+                    offset = 0;
+                }
+                let instance = unsafe { #type_ident::new(alloy_primitives::U256::from(next_slot), offset) };
+                offset += #type_ident::SLOT_BYTES as u8;
+                instance
+            }
+        };
+
+        field_initializations.push(quote! {
+            #field_name: #field_init
+        });
+    }
+
+    quote! {
+        use stylus_sdk::prelude::StorageType;
```

Comment:
A corollary of the previous change is that we no longer need to import this here, which means we can remove the modules in the tests.

```suggestion
```

Resulting in:

```rust
use alloy_primitives::{address, Address};
use alloy_primitives::{U128, U16, U256, U32, U64, U8};
use stylus_sdk::stylus_proc::sol_storage;

sol_storage! {
    #[derive(motsu_proc::DefaultStorageLayout)]
    pub struct Erc20 {
        mapping(address => uint256) _balances;
        mapping(address => mapping(address => uint256)) _allowances;
        uint256 _total_supply;
    }
}

#[motsu::test]
fn erc20_initiates_correctly(contract: Erc20) {
    assert_eq!(contract._total_supply.get(), U256::from(0));
}

sol_storage! {
    #[derive(motsu_proc::DefaultStorageLayout)]
    pub struct UintContract {
        uint256 a;
        uint128 b;
        uint64 c;
        uint32 d;
        uint16 e;
        uint8 f;
        uint8 g;
    }
}

// ...

#[motsu::test]
fn address_updates_correctly(contract: AddressContract) {
    let lender = address!("a935CEC3c5Ef99D7F1016674DEFd455Ef06776C5");
    let borrower = address!("DeaDbeefdEAdbeefdEadbEEFdeadbeEFdEaDbeeF");
    contract.lender.set(lender);
    contract.borrower.set(borrower);
    assert_eq!(contract.lender.get(), lender);
    assert_eq!(contract.borrower.get(), borrower);
}
```